lib/3scale/backend/stats/cleaner.rb

```diff
@@ -0,0 +1,156 @@
+module ThreeScale
+  module Backend
+    module Stats
+      class Cleaner
+        # Design notes:
+        # Apisonator does not store in any Redis structure the stats keys
+        # associated with a service. Doing so would imply:
+        #   - Performance hit when reporting. After updating a stats key, it
+        #   would need to be included in a set, which would increase the number
+        #   of operations in Redis.
+        #   - More space in Redis. To maintain the sets mentioned in the point
+        #   above.
+        #   - A data migration would be needed to create those sets from the
+        #   existing stats keys.
+        #
+        # In order to avoid those costs, this class is implemented in a way that
+        # does not need to keep an updated list of all the stats keys for every
+        # service. Services marked for deletion are stored in a Redis set, and
+        # then, a script, periodic cron, etc. is responsible for calling
+        # delete!(). That method scans the whole database and deletes all the
+        # stats keys that belong to those services marked to be deleted.
+        # The downside of this method is that it requires direct access to the
+        # redis servers. Going through a proxy like Twemproxy does not work,
+        # because it does not support the "SCAN" command.
+        #
+        # In the past we tried an alternative approach. When we received a
+        # request to delete the stats of a service, we generated all the
+        # possible stats keys that could exist for it. That approach was not
+        # efficient because it ended up generating many keys that didn't exist
+        # and thus, unnecessary delete calls to Redis. That approach is also
+        # more complex and error prone. You can find the details here:
+        # https://github.com/3scale/apisonator/issues/90
+
+        include Storable
+
+        SET_WITH_SERVICES_MARKED_FOR_DELETION = 'set_with_services_marked_for_deletion'.freeze
```

Comment:
Well, the actual name could be just `service_to_delete` or something shorter and simpler, since in the end _if_ you go inspecting the actual key you might as well know (or query) its type.

Reply:
Fixed :+1: